packages/pg/lib/crypto/crypto-shim.js

```diff
@@ -0,0 +1,101 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+const sha256Sync = require('./sha256')
+
+/**
+ * Patch the Node.js crypto APIs to work outside Node.js
+ */
+module.exports = function patchCrypto(crypto) {
+  crypto.randomBytes = function (length) {
+    return globalThis.crypto.getRandomValues(Buffer.alloc(length))
+  }
+
+  crypto.createHash = function (type) {
+    if (type !== 'sha256') throw new Error('Only sha256 is supported')
+    return {
+      update: function (data) {
+        return {
+          digest: function () {
+            return Buffer.from(sha256Sync(data))
+          },
+        }
+      },
+    }
+  }
+
+  crypto.createHmac = function createHmac(type, key) {
+    if (type !== 'sha256') throw new Error('Only sha256 is supported')
+    return {
+      update: function (data) {
+        return {
+          digest: function () {
+            if (typeof key === 'string') key = new TextEncoder().encode(key)
+            if (typeof data === 'string') data = new TextEncoder().encode(data)
+
+            const keyLen = key.length
+            if (keyLen > 64) {
+              key = sha256Sync(key)
+            } else if (keyLen < 64) {
+              const tmp = new Uint8Array(64)
+              tmp.set(key)
+              key = tmp
+            }
+            // else key length is 64, so leave well alone
+
+            const innerKey = new Uint8Array(64)
+            const outerKey = new Uint8Array(64)
+            for (let i = 0; i < 64; i++) {
+              innerKey[i] = 0x36 ^ key[i]
+              outerKey[i] = 0x5c ^ key[i]
+            }
+
+            const msg = new Uint8Array(data.length + 64)
+            msg.set(innerKey, 0)
+            msg.set(data, 64)
+
+            const result = new Uint8Array(64 + 32)
+            result.set(outerKey, 0)
+            result.set(sha256Sync(msg), 64)
+
+            return Buffer.from(sha256Sync(result))
+          },
+        }
+      },
+    }
+  }
+
+  crypto.pbkdf2Sync = function pbkdf2Sync(password, saltBytes, iterations, keylen, digest) {
```

Comment:
One limitation that Web Crypto does have here is that it cannot do PBKDF2 synchronously.

Reply:
Right, but as far as I know, we don’t need it to be synchronous (and it’s possible that it never should have been synchronous in the first place?).

Reply:
> (and it’s possible that it never should have been synchronous in the first place?).

yeah probably should have been async since blocking calls in a code path like this === bad

Reply:
Yep - I was trying to touch the original code as little as possible.
But now that this PR has shown this to be working I can look at further refactoring and improvements.